src/qudi/hardware/dummy/scanning_probe_dummy.py

```diff
@@ -583,24 +761,46 @@ def __stop_timer(self):
         else:
             self.__update_timer.stop()
 
-    def _init_position_vectors(self) -> Dict[str, np.ndarray]:
-        position_vectors = {}
-        x_axis = self.scan_settings.axes[0]
-        x_values = np.linspace(self.scan_settings.range[0][0],
-                               self.scan_settings.range[0][1],
-                               self.scan_settings.resolution[0])
-        position_vectors[x_axis] = x_values
-        if self.scan_settings.scan_dimension == 2:
-            y_axis = self.scan_settings.axes[1]
-            y_values = np.linspace(self.scan_settings.range[1][0],
-                                   self.scan_settings.range[1][1],
-                                   self.scan_settings.resolution[1])
-            position_vectors[y_axis] = y_values
-        return self._expand_coordinate(position_vectors)
+    def _init_scan_vectors_from_scan_settings(self) -> Dict[str, np.ndarray]:
+        axes_scan_values = [
+            np.linspace(
+                self.scan_settings.range[ii][0],
+                self.scan_settings.range[ii][1],
+                self.scan_settings.resolution[ii],
+            )
+            for ii, _ in enumerate(self.scan_settings.axes)
+        ]
+        # generate all combinations of points
+        meshgrids = np.meshgrid(*axes_scan_values, indexing="ij")
+        # create position vector dictionary by raveling the grids
+        scan_vectors = {
+            axis: grid.ravel() for axis, grid in zip(self.scan_settings.axes, meshgrids)
+        }
+        return scan_vectors
+
+    def _init_scan_vectors(self) -> Dict[str, np.ndarray]:
+        scan_vectors = self._init_scan_vectors_from_scan_settings()
+        scan_vectors = self._expand_coordinate(scan_vectors)  # always expand to all scan dims
+
+        return scan_vectors
+
+    @_spot_density.constructor
+    def spot_density_constructor(self, spot_density: float) -> float:
+        volume_edges = [abs(pos_range[1] - pos_range[0]) for pos_range in self._position_ranges.values()]
+        volume = 1
+        for edge in volume_edges:
+            volume *= edge
+        if volume * spot_density ** len(self._position_ranges.keys()) > 80e3:
```

Comment:
Maximum spot number should be a hardware constraint.

Reply:
I implemented it as a constraint and therefore subclassed the `ScanConstraint` class. This however results in the maximum spot number check to occur only during `on_activate` of the module and not during the constructor of the `ConfigOption`.